File: 11/b.go
package main

import (
	"fmt"
	"os"
	"slices"
	"strings"
)

type Star struct {
	x, y int
}

func main() {
	file, _ := os.ReadFile("input.txt")
	lines := strings.Split(string(file), "\n")

	emptyRows := make([]int, 0)
	emptyColumns := make([]int, len(lines[0]))
	for i := 0; i < len(lines[0]); i++ {
		emptyColumns[i] = i
	}
	stars := make([]Star, 0)
	for i, line := range lines {
		if line == "" {
			continue
		}
		isRowEmpty := true
		for j, char := range line {
			if char == '#' {
				isRowEmpty = false
				jIndex := slices.Index(emptyColumns, j)
				if jIndex != -1 {
					emptyColumns = slices.Delete(emptyColumns, jIndex, jIndex+1)
				}
				stars = append(stars, Star{x: j, y: i})
			}
		}
		if isRowEmpty {
			emptyRows = append(emptyRows, i)
		}
	}

	for i := 0; i < len(stars); i++ {
		addX := 0
		addY := 0
		for j := 0; j < len(emptyRows); j++ {
			if stars[i].y > emptyRows[j] {
				addY += 999999
			} else {
				break
			}
		}
		for j := 0; j < len(emptyColumns); j++ {
			if stars[i].x > emptyColumns[j] {
				addX += 999999
			} else {
				break
			}
		}
		stars[i].x += addX
		stars[i].y += addY
	}

	res := 0
	for i := 1; i < len(stars); i++ {
		for j := 0; j < i; j++ {
			dx := stars[i].x - stars[j].x
			dy := stars[i].y - stars[j].y
			if dx < 0 {
				dx = -dx
			}
			if dy < 0 {
				dy = -dy
			}
			res += dx + dy
		}
	}

	fmt.Println(res)
}

File: 11/b.ts
// const input = await Bun.file("test.txt").text();
const input = await Bun.file("input.txt").text();

let lines = input.split("\n").map((d) => d.split(""));
let freeColumns = lines[0].map((_, i) => i);
let freeRows = lines.map((_, i) => i);
lines.forEach((line, y) => {
  line.forEach((char, x) => {
    if (char === "#") {
      freeColumns = freeColumns.filter((c) => c !== x);
      freeRows = freeRows.filter((r) => r !== y);
    }
  });
});
const stars: [number, number][] = [];
let addY = 0;
lines.forEach((line, y) => {
  let addX = 0;
  if (freeRows.includes(y)) addY += 999_999;
  line.forEach((char, x) => {
    if (freeColumns.includes(x)) addX += 999_999;
    if (char === "#") {
      stars.push([x + addX, y + addY]);
    }
  });
});

let res = 0;

for (let i = 0; i < stars.length; i++) {
  const [x, y] = stars[i];
  for (let j = 0; j < i; j++) {
    const [x2, y2] = stars[j];
    const dx = Math.abs(x2 - x);
    const dy = Math.abs(y2 - y);
    const distance = dx + dy;
    res += distance;
  }
}

console.log(stars);
console.log(lines.map((d) => d.join("")).join("\n"));
console.log();
console.log(res);
